### PR TITLE
Test Refactor

### PR DESCRIPTION
**Pull request by redsn0w422**
#### Summary
Tests don't pass consistently due to race conditions, and this PR aims to fix those.

This PR:
- Refactors Unilog's `run` method to have a `tick`, so that we can test that it behaves correctly upon receiving different signals.
- Refactors Unilog's tests to remove race conditions.
- Refactors Reader to remove the race condition with the `shuttingDown` boolean by adding a mutex.
- Removes old (and probably broken) Reader code.
- Adds new Reader tests in `reader_test.go` and deprecates the old one.

TL;DR: running `go test -race` now passes!

#### Motivation
Currently, the tests in master don't pass consistently, which makes opening new PR's to Unilog difficult, since the existing tests may or may not (more common) pass.

#### Test plan
`go test` locally.

#### Rollout/monitoring/revert plan
Puppet the world

r? @asf-stripe 
cc @stripe/observability

### CODE REVIEW

**Comment by ChimeraCoder** [comment]
Good catch. We should add the race detector to the tests as well!

**Comment by redsn0w422** [comment]
@ChimeraCoder is that different than `go test -race`?

**Comment by ChimeraCoder** [comment]
Nope, that's it!

File: reader.go [hunk 1]
```diff
@@ -20,23 +24,27 @@ func NewUnilogReader(in io.Reader, shutdown <-chan struct{}) io.Reader {
...
-		return 0, io.EOF
-	}
+func (r *UnilogReader) isShuttingDown() bool {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	return r.shuttingDown
+}
 
-	if u.shuttingDown {
-		buf = buf[:1]
```
Comment (asf-stripe) [inline reader.go]: goodest of riddances to this case \o/ 🎉

File: reader_test.go [hunk 2]
```diff
@@ -78,3 +79,56 @@ func TestUnilogReader(t *testing.T) {
...
+func (or *OurReader) Read(p []byte) (n int, err error) {
+	bs, ok := <-or.input
+	if !ok {
+		return 0, io.EOF
+	}
+	return copy(p, bs), nil
+}
+
+func TestReaderAlone(t *testing.T) {
+	t.Run("basic", func(t *testing.T) {
```
Comment (asf-stripe) [inline reader_test.go]: I think you could `t.Parallel()` these (or make them separate test functions, really - they don't share any setup code...)

File: unilog.go [hunk 3]
```diff
@@ -224,31 +225,36 @@ func (u *Unilog) logLine(line string) {
 
 func (u *Unilog) run() {
 	for {
+		u.selector()
+	}
+}
+
+func (u *Unilog) selector() {
```
Comment (asf-stripe) [inline unilog.go]: Not loving this name... but can't think of something more descriptive at the moment. /:

File: reader_test.go (hunk 2 above)
Comment (redsn0w422) [inline reader_test.go]: Good point! I'll separate them.
